src/envelope.rs

```diff
@@ -305,13 +277,29 @@ impl<D: Hash> Envelope<D> {
             .zip(self.ciphertext.iter())
             .map(|(&x1, &x2)| x1 ^ x2)
             .collect();
-        Ok(OpenedEnvelope {
+        Ok(OpenedInnerEnvelope {
             plaintext,
             export_key: *GenericArray::from_slice(&export_key),
         })
     }
 }
 
+// Helper functions
+
+fn construct_aad(server_s_pk: &[u8], optional_ids: &Option<(Vec<u8>, Vec<u8>)>) -> Vec<u8> {
+    optional_ids
+        .iter()
+        .flat_map(|(l, r)| [serialize(server_s_pk, 2), serialize(l, 2), serialize(r, 2)].concat())
+        .collect()
+}
```

Comment:
I'm afraid this will need an adaptation, since we would still like to get `[serialize(server_s_pk, 2)]` when `optional_ids` is `None`. 

What I meant to suggest was :
```
let server_ids = optional_ids
        .iter()
        .flat_map(|(l, r)| [serialize(l, 2), serialize(r, 2)].concat())
        .collect();
[serialize(server_s_pk, 2), server_ids].concat()
 ```

Reply:
Oops, thanks for the catch! #112